core/storage/merkle/polkadot_trie_db/polkadot_codec.cpp

```diff
@@ -0,0 +1,222 @@
+    if (nibbles_size == 0) {
+      return {};
+    }
+
+    if (nibbles_size == 1 && nibbles[0] == 0) {
+      return {0};
+    }
+
+    // if nibbles_size is odd, then allocate one more item
+    const size_t size = (nibbles_size + 1) / 2;
+    common::Buffer out(size, 0);
+
+    size_t iterations = size;
+
+    // if number of nibbles is odd, then iterate even number of times
+    bool nimbles_size_odd = nibbles_size % 2 != 0;
+    if (nimbles_size_odd) {
+      --iterations;
+    }
+
+    for (size_t i = 0; i < iterations; ++i) {
+      out[i] = collectByte(nibbles[2 * i], nibbles[2 * i + 1]);
+    }
+
+    // if number of nibbles is odd, then implicitly add 0 as very last nibble
+    if (nimbles_size_odd) {
+      out[iterations] = collectByte(nibbles[2 * iterations], 0);
+    }
+
+    return out;
+  }
+
+  common::Buffer PolkadotCodec::keyToNibbles(const common::Buffer &key) const {
+    auto nibbles = key.size() * 2;
+
+    // if last nibble in `key` is 0
+    bool last_nibble_0 =
+        key.size() > 0 && (high4nibbles(key[key.size() - 1]) == 0);
+    if (last_nibble_0) {
+      --nibbles;
+    }
+
+    common::Buffer out(nibbles, 0);
+    for (size_t i = 0u, size = nibbles / 2; i < size; i++) {
+      out[2 * i] = low4nibbles(key[i]);
+      out[2 * i + 1] = high4nibbles(key[i]);
+    }
+
+    if (last_nibble_0) {
+      out[nibbles] = low4nibbles(key[key.size() - 1]);
+    }
+
+    return out;
+  }
+
+  common::Hash256 PolkadotCodec::hash256(const common::Buffer &buf) const {
+    common::Hash256 out;
+
+    static constexpr uint32_t k256 = 32;
+    if (buf.size() < k256) {
+      std::copy(buf.begin(), buf.end(), out.begin());
+      return out;
+    }
+
+    blake2s(out.data(), k256, nullptr, 0, buf.toBytes(), buf.size());
+    return out;
+  }
+
+  outcome::result<common::Buffer> PolkadotCodec::encodeNode(
+      const Node &node) const {
+    switch (static_cast<PolkadotNode::Type>(node.getType())) {
+      case PolkadotNode::Type::Branch:
+        return encodeBranch(dynamic_cast<const BranchNode &>(node));
+      case PolkadotNode::Type::Leaf:
+        return encodeLeaf(dynamic_cast<const LeafNode &>(node));
+    }
+
+    return std::errc::invalid_argument;
+  }
+
+  outcome::result<common::Buffer> PolkadotCodec::getHeader(
+      const PolkadotNode &node) const {
+    if (node.key_nibbles.size() > 0xffffu) {
+      return Error::TOO_MANY_NIBBLES;
+    }
+
+    uint8_t head = 0;
+    // set bits 6..7
+    switch (static_cast<PolkadotNode::Type>(node.getType())) {
+      case PolkadotNode::Type::Branch:
+        // has value?
+        head = node.value.size() > 0 ? 0b11 : 0b10;
+        break;
+      case PolkadotNode::Type::Leaf:
+        head = 0b01;
+        break;
+      default:
+        return Error::UNKNOWN_NODE_TYPE;
+    }
+    // FIXME: in the spec above two bits are most significant
+    //    head <<= 6u;  // head_{6..7} * 64
+
+    // set bits 0..5
+    if (node.key_nibbles.size() < 63u) {
```

Comment:
Are you sure that it is not like <64u ?
At the next line you moving it left by 2 bits, and "< 64" would mean that there will be no overflow.

Reply:
I am sure. This is how it is described in spec